File: testing-framework/backend/src/routes/export.ts
import { Router } from "express";
import { prisma } from "../prisma.js";
import { outlineTextToFreemind } from "../services/outlineToFreemind.js";

export const exportRouter = Router();

exportRouter.post("/xmind", async (req, res) => {
  const { requirementIds, testCaseIds, format } = req.body as {
    requirementIds?: string[];
    testCaseIds?: string[];
    format?: "txt" | "mm";
  };
  let cases: { caseId: string; featurePointL1: string; featurePoint: string; title: string; priority: string; preconditions: string; steps: string; expected: string; validationPoints: string; requirement: { title: string } }[];
  if (testCaseIds?.length) {
    const raw = await prisma.testCase.findMany({
      where: { id: { in: testCaseIds } },
      include: { requirement: { select: { title: true } } },
    });
    cases = raw;
  } else if (requirementIds?.length) {
    const raw = await prisma.testCase.findMany({
      where: { requirementId: { in: requirementIds } },
      include: { requirement: { select: { title: true } } },
    });
    cases = raw;
  } else {
    return res.status(400).json({ error: "Provide requirementIds or testCaseIds" });
  }
  /** 二级节点：优先用二级功能点，否则从标题取「xxx-」前部分 */
  function getL2(c: { featurePoint?: string | null; title: string }): string {
    if (c.featurePoint != null && String(c.featurePoint).trim()) return String(c.featurePoint).trim();
    const match = c.title.match(/^([^-－]+)[-－]/);
    return match ? match[1].trim() : "";
  }
  cases.sort(
    (a, b) =>
      a.requirement.title.localeCompare(b.requirement.title) ||
      (a.featurePointL1 ?? "").trim().localeCompare((b.featurePointL1 ?? "").trim()) ||
      getL2(a).localeCompare(getL2(b)) ||
      a.caseId.localeCompare(b.caseId)
  );
  const lines: string[] = [];
  let lastReq = "";
  let lastL1 = "";
  let lastL2 = "";
  for (const c of cases) {
    const reqTitle = c.requirement.title;
    const l1 = (c.featurePointL1 ?? "").trim();
    const l2 = getL2(c);
    if (reqTitle !== lastReq) {
      lines.push(reqTitle);
      lastReq = reqTitle;
      lastL1 = "";
      lastL2 = "";
    }
    if (l1 && l1 !== lastL1) {
      lines.push(`\t${l1}`);
      lastL1 = l1;
      lastL2 = "";
    }
    if (l2 && l2 !== lastL2) {
      lines.push(`\t\t${l2}`);
      lastL2 = l2;
    }
    const caseIndent = l1 || l2 ? (l1 && l2 ? "\t\t\t" : l1 ? "\t\t" : "\t") : "\t";
    const detailIndent = caseIndent + "\t";
    lines.push(`${caseIndent}${c.caseId} ${c.title}`);
    if (c.priority) lines.push(`${detailIndent}优先级：${c.priority}`);
    if (c.preconditions) lines.push(`${detailIndent}前置条件：${c.preconditions.replace(/\n/g, " ")}`);
    if (c.steps) lines.push(`${detailIndent}测试步骤：${c.steps.replace(/\n/g, " ")}`);
    if (c.expected) lines.push(`${detailIndent}预期结果：${c.expected.replace(/\n/g, " ")}`);
    if (c.validationPoints) {
      const vps = c.validationPoints.split(/\n/).filter(Boolean);
      vps.forEach((vp: string) => lines.push(`${detailIndent}验证点：${vp.replace(/\n/g, " ")}`));
    }
  }
  const outlineText = lines.join("\n");
  if (format === "mm") {
    const mm = outlineTextToFreemind(outlineText);
    res.setHeader("Content-Type", "application/xml; charset=utf-8");
    res.setHeader("Content-Disposition", "attachment; filename=xmind_outline.mm");
    res.send(Buffer.from(mm, "utf-8"));
    return;
  }
  res.setHeader("Content-Type", "text/plain; charset=utf-8");
  res.setHeader("Content-Disposition", "attachment; filename=xmind_outline.txt");
  res.send(outlineText);
});

File: testing-framework/backend_python/app/routers/export.py
from __future__ import annotations

import re
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.models import TestCase
from app.services.outline_to_freemind import outline_text_to_freemind

router = APIRouter(prefix="/export", tags=["export"])


def _get_l2(c: dict[str, Any]) -> str:
    fp = c.get("featurePoint")
    if fp is not None and str(fp).strip():
        return str(fp).strip()
    title = c.get("title") or ""
    m = re.match(r"^([^-－]+)[-－]", str(title))
    return m.group(1).strip() if m else ""


@router.post("/xmind")
def export_xmind(body: dict, db: Session = Depends(get_db)):
    requirement_ids = body.get("requirementIds") or []
    test_case_ids = body.get("testCaseIds") or []
    fmt = body.get("format") or "txt"
    if test_case_ids:
        raw = (
            db.query(TestCase)
            .options(joinedload(TestCase.requirement))
            .filter(TestCase.id.in_(test_case_ids))
            .all()
        )
    elif requirement_ids:
        raw = (
            db.query(TestCase)
            .options(joinedload(TestCase.requirement))
            .filter(TestCase.requirementId.in_(requirement_ids))
            .all()
        )
    else:
        raise HTTPException(status_code=400, detail="Provide requirementIds or testCaseIds")

    cases: list[dict[str, Any]] = []
    for c in raw:
        req = c.requirement
        cases.append(
            {
                "caseId": c.caseId,
                "featurePointL1": c.featurePointL1,
                "featurePoint": c.featurePoint,
                "title": c.title,
                "priority": c.priority,
                "preconditions": c.preconditions,
                "steps": c.steps,
                "expected": c.expected,
                "validationPoints": c.validationPoints,
                "requirement": {"title": req.title if req else ""},
            }
        )

    cases.sort(
        key=lambda a: (
            (a["requirement"]["title"] or ""),
            (str(a.get("featurePointL1") or "").strip()),
            _get_l2(a),
            a["caseId"],
        )
    )

    lines: list[str] = []
    last_req = ""
    last_l1 = ""
    last_l2 = ""
    for c in cases:
        req_title = c["requirement"]["title"]
        l1 = (c.get("featurePointL1") or "").strip()
        l2 = _get_l2(c)
        if req_title != last_req:
            lines.append(req_title)
            last_req = req_title
            last_l1 = ""
            last_l2 = ""
        if l1 and l1 != last_l1:
            lines.append(f"\t{l1}")
            last_l1 = l1
            last_l2 = ""
        if l2 and l2 != last_l2:
            lines.append(f"\t\t{l2}")
            last_l2 = l2
        case_indent = "\t\t\t" if (l1 and l2) else ("\t\t" if l1 else "\t")
        detail_indent = case_indent + "\t"
        lines.append(f"{case_indent}{c['caseId']} {c['title']}")
        if c.get("priority"):
            lines.append(f"{detail_indent}优先级：{c['priority']}")
        if c.get("preconditions"):
            pc = str(c["preconditions"]).replace("\n", " ")
            lines.append(f"{detail_indent}前置条件：{pc}")
        if c.get("steps"):
            st = str(c["steps"]).replace("\n", " ")
            lines.append(f"{detail_indent}测试步骤：{st}")
        if c.get("expected"):
            ex = str(c["expected"]).replace("\n", " ")
            lines.append(f"{detail_indent}预期结果：{ex}")
        if c.get("validationPoints"):
            vps = [x for x in str(c["validationPoints"]).split("\n") if x]
            for vp in vps:
                lines.append(f"{detail_indent}验证点：{vp.replace(chr(10), ' ')}")

    outline_text = "\n".join(lines)
    if fmt == "mm":
        mm = outline_text_to_freemind(outline_text)
        return Response(
            content=mm.encode("utf-8"),
            media_type="application/xml; charset=utf-8",
            headers={"Content-Disposition": 'attachment; filename="xmind_outline.mm"'},
        )
    return Response(
        content=outline_text.encode("utf-8"),
        media_type="text/plain; charset=utf-8",
        headers={"Content-Disposition": 'attachment; filename="xmind_outline.txt"'},
    )
